File: src/2023/week1/543.diameter_of_binary_tree.cpp
/**
 * Definition for a binary tree node.
 */
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

#include <algorithm>

class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        if (root == nullptr) return 0;

        int left = diameterOfBinaryTree(root->left);
        int right = diameterOfBinaryTree(root->right);
        int self = depth(root->left) + depth(root->right);
        int max_left_right = std::max(left, right);
        return std::max(max_left_right, self);
    }

    int depth(TreeNode *root) {
        if (root == nullptr) return 0;
        return depth(root->left) + depth(root->right) + 1;
    }
};

File: src/2023/week1/543.diameter_of_binary_tree.py
# Definition for a binary tree node.
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        return 0 if root is None else max(
            self.diameterOfBinaryTree(root.left),
            self.diameterOfBinaryTree(root.right),
            self.depth(root.left) + self.depth(root.right)
        )


    def depth(self, root: Optional[TreeNode]) -> int:
        if root is None:
            return 0
        return 0 if root is None else (
            max(self.depth(root.left), self.depth(root.right)) + 1
        )



